Initial implementation

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This pull request implements an Ansible role to configure [Transparent HugePage (THP)](https://www.kernel.org/doc/html/next/admin-guide/mm/transhuge.html) settings.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

The [GekoCloud/ansible-role-disable-thp](https://github.com/GekoCloud/ansible-role-disable-thp) role is a third-party dependency we use that is not actively maintained. I have had to make a pull request in the past to update that role to align with newer [ansible-lint](https://github.com/ansible/ansible-lint) rules. This has come up again and it made sense at this point to have our own role that implements this functionality to allow straightforward maintenance.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

Automated tests pass.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

mcdonnnj committed:
Create a service file to configure transparent hugepages

This SystemD service file is based on the work in
GekoCloud/ansible-role-disable-thp and the sample service file found at
https://www.mongodb.com/docs/manual/tutorial/transparent-huge-pages/#create-a-service-file
It allows configuration of the `defrag` and `enabled` settings through
sysfs.

mcdonnnj committed:
Create testing for the role

mcdonnnj committed:
Add a variable to configure a `Before` statement for the unit

This will allow configuration of the unit file to start before another
service.

mcdonnnj committed:
Adjust variables for the role

Make both the `manage_thp_enabled_setting` and
`manage_thp_defrag_setting` variables completely optional. This removes
default values for both variables and adjusts the check for both to
only be performed if the variable is defined.

mcdonnnj committed:
Update the role metadata

mcdonnnj committed:
Update the README

Replace skeleton default text with information specific to this role.

mcdonnnj committed:
Add support for two additional options

Allows configuration of the `shmem_enabled` and `use_zero_page`
options.

mcdonnnj committed:
Enhance testing for the role

In addition to verifying that the service is enabled we also check that
the unit file is valid using the `systemd-analyze` tool.

mcdonnnj committed:
Add fail logic to the role

If no Transparent HugePage setting is provided the role will fail. This
is done because if no settings are provided an invalid unit file would
be generated. A molecule scenario is created to test this and both the
GitHub Actions and pre-commit configurations are updated to support
this new scenario.

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

mcdonnnj committed:
Add a note about variables to the README

Mention that although none of the individual setting variables are
required, at least one of them must be defined.

dav3r commented on molecule/no_settings/converge.yml:
```diff
@@ -0,0 +1,19 @@
+              - ansible_failed_task.name == "Verify that at least one Transparent HugePage setting is provided"
+              - ansible_failed_result is defined
+              - ansible_failed_result.changed is false
+              - ansible_failed_result.failed is true
```
Nice test! 👍   Something like this would be great to incorporate into `skeleton-ansible-role`.

dav3r commented on README.md:
```diff
@@ -3,28 +3,33 @@
+| manage\_thp\_enabled\_setting | A value of `always`, `madvise`, or `never`. Please see the [kernel documentation] for more information. | n/a | No |
+| manage\_thp\_service\_name | The name of the SystemD service that is created to manage Transparent HugePage settings. Please note that `.service` is appended to this value in the name of the unit file created. | `configure-transparent-hugepage-settings` | No |
+| manage\_thp\_shmem\_enabled\_path | The sysfs path to control the Transparent HugePage `shmem_enabled` setting. | `/sys/kernel/mm/transparent_hugepage/shmem_enabled` | No |
+| manage\_thp\_shmem\_enabled\_setting | If defined it must be a value of `always`, `within_size`, `advise`, `never`, `deny`, or `force`. Please see the [kernel documentation] for more information. | n/a | No |
```
Alphabetize?
```suggestion
| manage\_thp\_shmem\_enabled\_setting | If defined it must be a value of `advise`, `always`, `deny`, `force`, `never`, or `within_size`. Please see the [kernel documentation] for more information. | n/a | No |
```

jsf9k replied:
If @mcdonnnj is copying from the Linux documentation then I recommend retaining the same order used there.

mcdonnnj replied:
The sorting for all of these is based on output like so:

```console
$ cat /sys/kernel/mm/transparent_hugepage/shmem_enabled
always within_size advise [never] deny force
```

I elected to duplicate the ordering to align with how it is presented if done manually. Looking at the kernel documentation linked the ordering they have appears to be: `always`, `never`, `within_size`, `advise`, `deny`, `force`

jsf9k replied:
I'm fine with the ordering as it is.  I'm also fine if you want to alphabetize it.

dav3r replied:
Either is fine with me.  If we don't alphabetize, can we put a comment somewhere in here so that we don't revisit this in the future.

mcdonnnj replied:
@dav3r @jsf9k would you take a look at [259b4a3](https://github.com/cisagov/ansible-role-manage-thp/pull/4/commits/259b4a32ff6962c2d6e2894687b377fb79e39774) and see what that does for you?

dav3r replied:
That comment works for me, thanks!  I'm resolving this conversation.  @jsf9k feel free to reopen if you want any adjustments.

jsf9k replied:
Works for me too!

dav3r commented on tasks/main.yml:
```diff
@@ -1,2 +1,51 @@
+  ansible.builtin.assert:
+    fail_msg: The provided Transparent HugePage shmem_enabled setting is not valid
+    that:
+      - manage_thp_shmem_enabled_setting in ['always', 'within_size', 'advise', 'never', 'deny', 'force']
```
Alphabetize?
```suggestion
      - manage_thp_shmem_enabled_setting in ['advise', 'always', 'deny', 'force', 'never', 'within_size']
```

jsf9k replied:
If @mcdonnnj is copying from the Linux documentation then I recommend retaining the same order used there.

dav3r replied:
Resolving this conversation - please see https://github.com/cisagov/ansible-role-manage-thp/pull/4#discussion_r1578078546 for details.

dav3r reviewed (approved):
This looks great to me, nice work!  💪 💼 
I could only come up with tiny alphabetization quibbles.

jsf9k commented on README.md:
```diff
@@ -3,28 +3,33 @@
-| required_variable | Describe its purpose. | n/a | Yes |
--->
+| manage\_thp\_defrag\_path | The sysfs path to control the Transparent HugePage `defrag` setting. | `/sys/kernel/mm/transparent_hugepage/defrag` | No |
+| manage\_thp\_defrag\_setting | If defined it must be a value of `always`, `defer`, `defer+madvise`, `madvise`, or `never`. Please see the [kernel documentation] for more information. | n/a | No |
```
```suggestion
| manage\_thp\_defrag\_setting | If defined this variable must be given a value of `always`, `defer`, `defer+madvise`, `madvise`, or `never`. Please see the [kernel documentation] for more information. | n/a | No |
```

jsf9k commented on README.md:
```diff
@@ -3,28 +3,33 @@
+| manage\_thp\_defrag\_path | The sysfs path to control the Transparent HugePage `defrag` setting. | `/sys/kernel/mm/transparent_hugepage/defrag` | No |
+| manage\_thp\_defrag\_setting | If defined it must be a value of `always`, `defer`, `defer+madvise`, `madvise`, or `never`. Please see the [kernel documentation] for more information. | n/a | No |
+| manage\_thp\_enabled\_path | The sysfs path to control the Transparent HugePage `enabled` setting. | `/sys/kernel/mm/transparent_hugepage/enabled` | No |
+| manage\_thp\_enabled\_setting | A value of `always`, `madvise`, or `never`. Please see the [kernel documentation] for more information. | n/a | No |
```
```suggestion
| manage\_thp\_enabled\_setting | If defined this variable must be given a value of `always`, `madvise`, or `never`. Please see the [kernel documentation] for more information. | n/a | No |
```

jsf9k commented on README.md:
```diff
@@ -3,28 +3,33 @@
+| manage\_thp\_shmem\_enabled\_setting | If defined it must be a value of `always`, `within_size`, `advise`, `never`, `deny`, or `force`. Please see the [kernel documentation] for more information. | n/a | No |
+| manage\_thp\_start\_before\_service | The name of the service that the SystemD unit file should be configured to start before. Please note that `.service` is appended to this value in the definition of the `Before` directive. | n/a | No |
+| manage\_thp\_use\_zero\_page\_path | The sysfs path to control the Transparent HugePage `use_zero_page` setting. | `/sys/kernel/mm/transparent_hugepage/use_zero_page` | No |
+| manage\_thp\_use\_zero\_page\_setting | If defined it must be a value of `0` or `1`. Please see the [kernel documentation] for more information. | n/a | No |
```
```suggestion
| manage\_thp\_use\_zero\_page\_setting | If defined this variable must be given a value of `0` or `1`. Please see the [kernel documentation] for more information. | n/a | No |
```

jsf9k reviewed (changes requested):
Nicely done!  I'm requesting a few minor language changes, but nothing major.

mcdonnnj committed:
Improve variable description language

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

mcdonnnj committed:
Add comments explaining why we are not using alphabetical order

When listing out valid values from settings they are not always listed
in alphabetical order. This is intentionally done so that it aligns
with what one sees when checking the current configuration of a setting
on a live system. Since we are deviating from our normal practices it
is important that we document the decision and include the rationale.

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>
Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

jsf9k reviewed (approved):
_Approval intensifies!!!_

jsf9k commented:
@mcdonnnj - Can this PR be merged now?
